contracts/Strategy.sol

```diff
@@ -53,80 +222,624 @@ contract Strategy is BaseStrategy {
+                        )
+                    );
+                }
+
+                bentoBox.withdraw(
+                    BIERC20(address(want)),
+                    address(this),
+                    address(this),
+                    0,
+                    sharesInBento()
+                );
+            }
+
+            _liquidatedAmount = balanceOfWant();
+
+            if (_amountNeeded > _liquidatedAmount) {
+                _loss = _amountNeeded.sub(_liquidatedAmount);
+            }
         } else {
             _liquidatedAmount = _amountNeeded;
         }
     }
 
-    function liquidateAllPositions() internal override returns (uint256) {
-        // TODO: Liquidate all positions and return the amount freed.
+    function liquidateAllPositions()
+        internal
+        override
+        returns (uint256 _liquidatedAmount)
+    {
+        (_liquidatedAmount, ) = liquidatePosition(type(uint256).max);
+    }
+
+    // new strategy **must** have the same kashiPairs attached
+    function prepareMigration(address _newStrategy) internal override {
+        for (uint256 i = 0; i < kashiPairs.length; i++) {
+            if (kashiPairs[i].pid != 0) {
+                masterChef.withdraw(
+                    kashiPairs[i].pid,
+                    kashiFactionInMasterChef(i)
+                );
+            }
+
+            kashiPairs[i].kashiPair.transfer(
+                _newStrategy,
+                kashiFractionInPair(i)
+            );
+        }
+    }
+
+    function addKashiPair(address _newKashiPair, uint256 _newPid)
+        external
+        onlyGovernance
+    {
+        // cannot exceed max pair length
+        require(kashiPairs.length < MAX_PAIRS);
+        // must use the correct bentobox
+        require(
+            address(IKashiPair(_newKashiPair).bentoBox()) == address(bentoBox)
+        );
+        // kashPair asset must match want
+        require(IKashiPair(_newKashiPair).asset() == BIERC20(address(want)));
+        if (_newPid != 0) {
+            // masterChef pid token must match the kashiPair
+            require(
+                address(masterChef.poolInfo(_newPid).lpToken) == _newKashiPair
+            );
+        }
+
+        for (uint256 i = 0; i < kashiPairs.length; i++) {
+            // kashiPair must not already be attached
+            require(_newKashiPair != address(kashiPairs[i].kashiPair));
+        }
+
+        kashiPairs.push(KashiPairInfo(IKashiPair(_newKashiPair), _newPid));
+
+        if (_newPid != 0) {
+            IERC20(_newKashiPair).safeApprove(
+                address(masterChef),
+                type(uint256).max
+            );
+        }
+    }
+
+    function removeKashiPair(address _remKashiPair)
+        external
+        onlyEmergencyAuthorized
+    {
+        for (uint256 i = 0; i < kashiPairs.length; i++) {
+            if (_remKashiPair != address(kashiPairs[i].kashiPair)) continue;
+            liquidateKashiPair(
```

Comment:
i feel a few of these helper methods that handle the index, are rereading from storage the same address where we can just pass along the address and call the KashiPair interface directly.

Reply:
Sounds good